extended proxy authenticator to pass additional attributes via header

jcantrill opened the pull request:
This PR introduces and extended proxy authenticator to allow proxies to pass additional user attributes that can be consumed when using document level security.  Configured like:
```
    http:
      xff:
        enabled: true
        remoteIpHeader: 'x-forwarded-for'
        trustedProxies: '.*'
        internalProxies: '.*'
    authc:
      authentication_domain_proxy:
        enabled: true
        order: 0
        http_authenticator:
          challenge: false
          type: extended-proxy
          config:
            user_header: 'x-proxy-remote-user'
            attr_header_prefix: 'x-proxy-ext-'
        authentication_backend:
          type: noop
```
This allows a trusted proxy to provide request headers:
```
x-proxy-ext-namespace: foo
x-proxy-ext-namespace: bar
x-proxy-ext-bu: makemoney
```
that are made available as user attributes:
```
attr.proxy.namespace: foo,bar
attr.proxy.bu: makemoney
```

richm commented:
@alolita ptal

alolita commented:
Hi @richm will follow up - apologies for the delayed response.

hardik-k-shah commented:
Sorry for late reply.
We are looking into this now. 
@allenyin96 is creating required setup for testing.  

Meanwhile, do you mind sharing your proxy setup as well? 
Also It will be great if you can merge mainline and run UTs for security as well as security-advanced-module and update the results on this PR.

allenyin96 commented:
Hi, 

So seems like you guys wrote this against Open Distro 0.9, which was shipped with Elasticsearch 6.7.1. Our master branch currently contains Open Distro 1.2, which works with Elasticsearch 7.2.0. 

Some ES function has changed, i.e. `RestRequest` has changed its signature to [this](https://github.com/elastic/elasticsearch/blob/77a8bc3cd1ec585873daad50ed34577184204571/server/src/main/java/org/elasticsearch/rest/RestRequest.java#L76). 

If you still want this PR in Open Distro 0.9, please change the destination branch of this PR to  opendistro-0.9. 

Otherwise, please update your code so that it works with Open Distro 1.2.0 (ES 7.2.0) if you want this to be in our master (Open Distro 1.x) as well.

Thanks!

jcantrill commented:
> Hi,
> 
> So seems like you guys wrote this against Open Distro 0.9, which was shipped with Elasticsearch 6.7.1. Our master branch currently contains Open Distro 1.2, which works with Elasticsearch 7.2.0.
> 
Any chance there is going to be 6.8.x compatability?

> If you still want this PR in Open Distro 0.9, please change the destination branch of this PR to opendistro-0.9.
> 
> Otherwise, please update your code so that it works with Open Distro 1.2.0 (ES 7.2.0) if you want this to be in our master (Open Distro 1.x) as well.
> 
> Thanks!

allenyin96 commented:
Have you built the code into a jar and install it on an ES cluster to test? I tried building your code and installing it, and I had to add the following line to [here](https://github.com/opendistro-for-elasticsearch/security/blob/2adee3f258795bc7bf3a82d4846721021eb93010/src/main/java/com/amazon/opendistroforelasticsearch/security/auth/BackendRegistry.java#L215) to get ES to run.

`authImplMap.put("extended-proxy_h", HTTPProxyAuthenticator.class.getName());`

Also have you run the entire unit test suite?

jcantrill commented:
> @jcantrill Have you built the artifact and install it on an ES cluster to test? 

Not recently

I tried building your code and installing it, and I had to add the following line to [here](https://github.com/opendistro-for-elasticsearch/security/blob/2adee3f258795bc7bf3a82d4846721021eb93010/src/main/java/com/amazon/opendistroforelasticsearch/security/auth/BackendRegistry.java#L215) to get ES to run.
> 
> `authImplMap.put("extended-proxy_h", HTTPProxyAuthenticator.class.getName());`

Fixed to ref the extended authenticator class

> 
> Also have you run the entire unit test suite?

yes

> 
> You can create a PR to opendistro-0.10 branch if you want 6.8 compatibility.

Rebased to opendistro-0.10

allenyin96 commented:
Tested and worked.

Merging.

jcantrill commented:
@allenyin96 @hardik-k-shah Is it possible to get a 0.10 release with this commit?
